OpenRA.Mods.Common/Activities/DeliverUnit.cs

```diff
@@ -83,110 +65,69 @@ public override Activity Tick(Actor self)
+				QueueChild(self, new Wait(25));
+				return this;
+			}
 
-				case DeliveryState.Land:
+			// Move to drop-off location
+			var targetPosition = self.World.Map.CenterOfCell(targetLocation.Value);
+			var localOffset = carryall.CarryableOffset.Rotate(body.QuantizeOrientation(self, self.Orientation));
+			var carryablePosition = self.CenterPosition + body.LocalToWorld(localOffset);
+			if ((carryablePosition - targetPosition).HorizontalLengthSquared != 0)
+			{
+				// For non-zero offsets the drop position depends on the carryall facing
+				// We therefore need to predict/correct for the facing *at the drop point*
+				if (carryall.CarryableOffset.HorizontalLengthSquared != 0)
 				{
-					if (!CanDropHere())
-					{
-						state = DeliveryState.Transport;
-						return this;
-					}
-
-					// Make sure that the carried actor is on the ground before releasing it
-					var localOffset = carryall.CarryableOffset.Rotate(body.QuantizeOrientation(self, self.Orientation));
-					var carryablePosition = self.CenterPosition + body.LocalToWorld(localOffset);
-					if (self.World.Map.DistanceAboveTerrain(carryablePosition) != WDist.Zero)
-					{
-						QueueChild(self, new HeliLand(self, false, -new WDist(carryall.CarryableOffset.Z)), true);
-						return this;
-					}
-
-					state = carryall.Info.UnloadingDelay > 0 ? DeliveryState.Wait : DeliveryState.Release;
+					var facing = (targetPosition - self.CenterPosition).Yaw.Facing;
+					localOffset = carryall.CarryableOffset.Rotate(body.QuantizeOrientation(self, WRot.FromFacing(facing)));
+					QueueChild(self, new HeliFly(self, Target.FromPos(targetPosition - body.LocalToWorld(localOffset))), true);
+					QueueChild(self, new Turn(self, facing));
 					return this;
 				}
 
-				case DeliveryState.Wait:
-					state = DeliveryState.Release;
-					QueueChild(self, new Wait(carryall.Info.UnloadingDelay, false), true);
-					return this;
-
-				case DeliveryState.Release:
-					if (!CanDropHere())
-					{
-						state = DeliveryState.Transport;
-						return this;
-					}
-
-					Release();
-					state = DeliveryState.TakeOff;
-					return this;
+				QueueChild(self, new HeliFly(self, Target.FromPos(targetPosition)), true);
+				return this;
+			}
 
-				case DeliveryState.TakeOff:
-					QueueChild(self, new HeliFly(self, Target.FromPos(self.CenterPosition)), true);
-					state = DeliveryState.Done;
-					return this;
+			// Make sure that the carried actor is on the ground before releasing it
+			if (self.World.Map.DistanceAboveTerrain(carryablePosition) != WDist.Zero)
+				QueueChild(self, new HeliLand(self, true), true);
 
-				case DeliveryState.Aborted:
-					carryall.UnreserveCarryable(self);
-					break;
-			}
+			// Pause briefly before releasing for visual effect
+			if (carryall.Info.UnloadingDelay > 0)
+				QueueChild(self, new Wait(carryall.Info.UnloadingDelay, false), true);
 
-			return NextActivity;
+			// Release carried actor
+			QueueChild(self, new CallFunc(Release));
```

Comment:
We are working slowly towards removing `CallFunc` from the code base.
Would you mind splitting `Release` into a child activity (something like f79bba725ffb315e58fbb0c1996fa155dbbb5fb4) if we don't want to keep the switch case here?

More generally / alternatively: why get rid of the switch here, but keep it `PickupUnit`?

Reply:
Cherry-picked and slightly adjusted.

I kept the switch in `PickupUnit` because this PR was supposed to be about drop-off behaviour only. I intend to get rid of the switch there as well, but #16509 provides a better jump-off point for that.